File: agent_pay_for_urself/api/routes/decisions.py
"""Decision workflow route."""

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException

from agent_pay_for_urself.api.dependencies import get_decision_workflow_service
from agent_pay_for_urself.api.mappers.workflow import to_decision_response
from agent_pay_for_urself.api.models import DecisionRequest, DecisionResponse
from agent_pay_for_urself.api.services.decision_workflow import DecisionWorkflowService
from agent_pay_for_urself.schemas import InvestmentMandate

router = APIRouter(tags=["decisions"])
WorkflowService = Annotated[DecisionWorkflowService, Depends(get_decision_workflow_service)]


@router.post(
    "/decisions",
    response_model=DecisionResponse,
    summary="Run investment decisions",
    description=(
        "Runs the current multi-agent workflow for the requested symbols: data collection, "
        "signal analysis, risk validation, buy/sell/hold decision, order planning, and "
        "evaluation summary. Use this endpoint to inspect analysis scores, risk checks, "
        "decisions, and non-submitted order plans even when a broker adapter is configured."
    ),
)
def create_decision(
    request: DecisionRequest,
    workflow_service: WorkflowService,
) -> DecisionResponse:
    """Run the orchestrated decision workflow and return a stored result."""

    try:
        run_id, result = workflow_service.run(
            symbols=request.symbols,
            max_position_weight=request.max_position_weight,
            mandate=_to_investment_mandate(request),
        )
        return to_decision_response(run_id, result, workflow_service.runtime_summary())
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(status_code=502, detail=str(exc)) from exc
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"decision workflow failed: {exc}") from exc


def _to_investment_mandate(request: DecisionRequest) -> InvestmentMandate | None:
    if request.mandate is None:
        return None

    return InvestmentMandate(
        objective=request.mandate.objective,
        allowed_symbols=tuple(request.mandate.allowed_symbols),
        excluded_symbols=tuple(request.mandate.excluded_symbols),
        max_position_weight=request.max_position_weight,
        max_order_notional=request.mandate.max_order_notional,
        min_cash_weight=request.mandate.min_cash_weight,
        risk_tolerance=request.mandate.risk_tolerance,
        requires_approval_for_live_orders=request.mandate.requires_approval_for_live_orders,
        user_notes=request.mandate.user_notes,
    )

File: artifacts/api-server/src/routes/decisions.ts
import { Router, type IRouter } from "express";
import { runWorkflow } from "../engine/orchestrator";
import type { InvestmentMandate, RiskTolerance } from "../engine/schemas";

const router: IRouter = Router();

router.post("/decisions", (req, res) => {
  const body = req.body as Record<string, unknown>;

  const symbols: string[] = Array.isArray(body.symbols)
    ? (body.symbols as unknown[]).map(String).filter(Boolean)
    : [];

  const maxPositionWeight =
    typeof body.max_position_weight === "number" ? body.max_position_weight : 0.2;

  const userPrompt = typeof body.user_prompt === "string" ? body.user_prompt : "";
  const chatMessages: string[] = Array.isArray(body.chat_messages)
    ? (body.chat_messages as unknown[]).map(String)
    : [];

  let mandate: InvestmentMandate | null = null;
  if (body.mandate && typeof body.mandate === "object") {
    const m = body.mandate as Record<string, unknown>;
    mandate = {
      objective:
        typeof m.objective === "string"
          ? m.objective
          : userPrompt || "Evaluate requested equity symbols conservatively.",
      allowed_symbols: Array.isArray(m.allowed_symbols)
        ? (m.allowed_symbols as unknown[]).map(String)
        : [],
      excluded_symbols: Array.isArray(m.excluded_symbols)
        ? (m.excluded_symbols as unknown[]).map(String)
        : [],
      max_position_weight:
        typeof m.max_position_weight === "number" ? m.max_position_weight : maxPositionWeight,
      max_order_notional:
        typeof m.max_order_notional === "number" ? m.max_order_notional : null,
      min_cash_weight:
        typeof m.min_cash_weight === "number" ? m.min_cash_weight : null,
      risk_tolerance:
        typeof m.risk_tolerance === "string"
          ? (m.risk_tolerance as RiskTolerance)
          : "medium",
      requires_approval_for_live_orders:
        typeof m.requires_approval_for_live_orders === "boolean"
          ? m.requires_approval_for_live_orders
          : true,
      user_notes: typeof m.user_notes === "string" ? m.user_notes : "",
    };
  }

  const promptOverrides: Record<string, string> = {};
  if (body.prompt_overrides && typeof body.prompt_overrides === "object") {
    const overrides = body.prompt_overrides as Record<string, unknown>;
    for (const [k, v] of Object.entries(overrides)) {
      if (typeof v === "string") promptOverrides[k] = v;
    }
  }

  if (symbols.length === 0 && !userPrompt.trim()) {
    res.status(422).json({ error: "symbols 또는 user_prompt 중 하나 이상을 입력하세요." });
    return;
  }

  const result = runWorkflow({
    symbols,
    max_position_weight: maxPositionWeight,
    mandate,
    user_prompt: userPrompt,
    chat_messages: chatMessages,
    prompt_overrides: promptOverrides,
  });

  res.json(result);
});

export default router;
